### DISCUSSION TITLE
XPRESS support in MathOpt

### DISCUSSION THREAD

**Discussion by pet-mit** [google/or-tools#4455] [Linear Solver questions]
Hello everyone,

We ([RTE](https://github.com/rte-france/or-tools)) have started the development of the "XpressSolver" interface to allow the use of MathOpt with the [FICO XPRESS solver](https://www.fico.com/en/products/fico-xpress-optimization).
The development will probably take a few months, and we'll implement features gradually (starting with LP, MIP, etc). 

We created this discussion so that we can have a place to chat about this topic.

(People from RTE involved in the developments: @pet-mit , @sgatto , @flomnes , @klorel )
(For your information: @djunglas )

**Comment by pet-mit** [comment]
Question: Would you prefer that we open several small pull requests to merge these features into the main branch one at a time, or would you prefer that we merge everything at once?
(@lperron @Mizux)

**Comment by rma350** [comment]
There are a few key differences in adding a solver for mathopt:

(1) The expectations of each solver on the input and output are higher (solver indepenent parameters, solver independent termination reasons) and there are integration tests that check that the behavior of the solvers is consistent. It is generally somewhat challening to get these tests working, but in the end it lets you do more complex things where you need these features (e.g. decomposition) in a solver indepent way. The tests can be a little brittle (the solvers are not actually the same/do not really offer strong contracts on there behavior). The tests need to be working from the start on whatever features you have. You may need to ask for help on getting everything working. Sometimes we need to change the tests.

(2) We try not to program directly against the C API, as it ends up being very verbose, and the code already is quite long. See gurobi (or the pull request for mosek) as good examples of wrapping the C API with an ABSL based C++ API (using Status for error propagation, and absl::span<const double> instead of double*, etc.  Try and keep the code as close to the C API as possible/document any behavior changes if we really need to do them. And test this file directly (just make sure each codepath is covered by the tests, you do not need to test xpress itself.)

(3) In some situations, there is not a 1:1 mapping between the mathopt model and the underlying solver model. E.g. with Gurobi, they do not natively support linear constraints with upper and lower bounds (range constraints), you need to add an auxiliary variable to model this. These model transformations, combined with (a) incrementalism (b) variable deletion and (c) duality, ends up adding quite a bit of complexity. To get started, I would recommend just not supporting incrementalism and solving from scratch each time, and most of these issues go away. If things eventually get too complicated with the model transformations, you can break them out into a separate file and unit test the transformations directly.

I would recommend starting with something as small as possible, and just trying to get the integration tests to pass, e.g. just LP (maybe no basis, no rays, minimial/no solver parameters). You can look at the glop tests for an example:

https://github.com/google/or-tools/blob/stable/ortools/math_opt/solvers/glop_solver_test.cc

(just focus on the LP tests, you can ignore the rest to get started). Internally, we use bazel for running the tests. I am not sure if the tests work with cmake yet. If you want the mathopt team to be able to reproduce any problems you are having, the team is generally more familiar with bazel (lperron and mizux do more with both bazel and cmake).

I think that any code which is in a clean state (everything tested, nothing commented out) can probably be merged. I do not know which branch you should be on, I guess lperron and mizux will. Working in smaller pieces will probalby result in faster reviews. I think we need to have a quick discussion internally about merging this/who will be responsible, it will need to wait until after Thanksgiving (next week) as too many people are out.

**Reply by pet-mit** [reply]
Thank you for your very detailed answer.

I have spent a few hours familiarizing myself with the MathOpt code, especially with the Gurobi interface.
In fact, I tried to replicate the same design choices you made for Gurobi to implement XPRESS (although I think I can simplify a few things, like the constraint creation that I replicated, but that can be simplified for XPRESS).

I have enabled the LP simple tests for XPRESS and am working on getting them to pass (they do compile with CMake).

However, I have not yet created a bazel file for the XPRESS wrapper (g_xpress) and would not know how to do so without your help.

I have opened a draft PR in our fork here. It's pretty simple if you want to take a look already (the simplest of LP tests pass currently): https://github.com/rte-france/or-tools/pull/137 

I have taken note of all your points, we will open small pull requests with the cleanest possible code to implement features one by one (under the condition that their tests pass, of course).

**Reply by rma350** [reply]
I looked through your change, I think this is all going on the correct direction. The Gurobi code is a good reference, but in particular, gurobi_solver.h/cc is a bit complex, you can/should probably do something simpler if possible.

I think the bazel part will be straight forward as well. We can figure it out once the code is working.

**Reply by pet-mit** [reply]
Thank you Ross for your valuable feedback

**Comment by pet-mit** [comment]
Hello
I'm struggling to make the test here pass for XPRESS, when using the primal simplex algorithm:
https://github.com/google/or-tools/blob/82914bc4f9490722acdce4e82adac0f8e11ff967/ortools/math_opt/solver_tests/lp_incomplete_solve_tests.cc#L111

How do we know for sure that there is a dual solution available (can we even know if dual is feasible?) when solving using primal simplex, if we interrupt it before the end?


@djunglas I don't know if you can help here: XPRESS seems to say that it has no dual solution in this case after interruption (XPRS_SOLSTATUS_NOTFOUND), when I use [XPRSgetduals](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/XPRSgetduals.html) to get the dual status

**Reply by rma350** [reply]
The solution will generally not be dual feasible. But if you have a basis, then there is still a canonical dual solution to write down. E.g. for standard form problems (min cx s.t. Ax = b, x>=0), if B is your set of basic variables and (A_B^T)-1 c_B is the dual infeasilbe solution. When you have variable upper and lower bounds everything is a big mess, but the idea is the same.

Likewise with dual simplex, you maintain a dual feasible solution throughout, but since you are always at a basis, you can also compute a primal infeasible solution for this basis.

It may be the case that XPRESS does not support extracting the dual solution. Then we could either disable the test for xpress or parameterize the test to potentially not expect a dual solution. But simplest would be to just return the dual. We do not want to recomptue the dual in MathOpt code (or in general, solve linear systems), as these operations can fail, and then we need to explain to the user that the failure was in our wrapper, not in the solver, which is confusing.

**Reply by pet-mit** [reply]
Thanks Ross, noted.
I will then produce a vector of dual values (as provided by the relevant XPRESS function) without setting dual feasibility to "true".
I agree that MathOpt mustn't recompute these values.

(FYI I still need to do some cleanup and will open the first PR next week!)

**Reply by rma350** [reply]
I think it is likely that when running dual simplex, you will be in the reversed situation if you stop early, your solution is dual feasible but not primal feasible.

Also in both cases, if you stop in "Phase 1" (before finding a primal or dual feasible soltuion), you may not be able to return anything. I do not know if xpress lets you get at the phase 1 solution/basis (I don't think all solvers let you do this, but I do not remember the details from when we did this for GLOP or Gurobi).

**Reply by pet-mit** [reply]
I verified this case, XPRESS does return a primal solution

**Reply by djunglas** [reply]
@pet-mit, as far as I can tell, the failing test only checks that there is _any_ dual value. It does not check that the dual values are correct or even reasonable, right?
I am wondering whether values returned by other solvers (that pass this test) are reasonable or whether these are just "some" values that happened to be in internal vectors at that time.

**Reply by pet-mit** [reply]
@djunglas you are right

FYI I have managed to make the test pass by providing the dual values returned by [XPRSgetlpsol](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/XPRSgetlpsol.html)
I opened the first PR here: https://github.com/google/or-tools/pull/4505

**Reply by djunglas** [reply]
The situation with Xpress is a bit tricky:
- For an interrupted primal simplex run, functions like `XPRSgetduals()` or `XPRSgetsolution()` will always return `XPRS_SOLSTATUS_NOTFOUND` (0)
- In this case you can use `XPRSgetlpsol()` (just like you did) to get the current values (you have to know what you are doing and which values you can expect to be valid)
- In all other cases you should use `XPRSgetduals()` and friends.

**Comment by pet-mit** [comment]
@djunglas 
On a seperate topic, I have noticed that when the primal is unbounded, [LPSTATUS](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/LPSTATUS.html) is equal to "XPRS_LP_UNBOUNDED", but the dual "status" fetched through [XPRSgetduals](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/XPRSgetduals.html) is equal to "[XPRS_SOLSTATUS_UNBOUNDED](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/SOLSTATUS.html)".  

Is this the expected behavior, or should the dual status be "XPRS_SOLSTATUS_INFEASIBLE"?

**Reply by djunglas** [reply]
This is a known issue and we are working on it. At the moment it is best to check the status returned by XPRSoptimize(), or, when using XPRSlpoptimize() to check the LPSTATUS.
